discord/appinfo.py

```diff
@@ -514,3 +520,188 @@ class AppInstallParams:
+        using :meth:`~discord.Client.fetch_emoji` or :meth:`~discord.Client.fetch_emojis`
+    """
+
+    __slots__: Tuple[str, ...] = (
+        'application_id',
+        'require_colons',
+        'animated',
+        'managed',
+        'id',
+        'name',
+        '_roles',
+        '_state',
+        'user',
+        'available',
+    )
+
+    def __init__(self, *, application_id: int, state: ConnectionState, data: EmojiPayload) -> None:
+        self.application_id: int = application_id
+        self._state: ConnectionState = state
+        self._from_data(data)
+
+    def _from_data(self, emoji: EmojiPayload) -> None:
+        from .user import User  # circular import
+
+        self.require_colons: bool = emoji.get('require_colons', False)
+        self.managed: bool = emoji.get('managed', False)
+        self.id: int = int(emoji['id'])  # type: ignore # This won't be None for full emoji objects.
+        self.name: str = emoji['name']  # type: ignore # This won't be None for full emoji objects.
+        self.animated: bool = emoji.get('animated', False)
+        self.available: bool = emoji.get('available', True)
+        self._roles: utils.SnowflakeList = utils.SnowflakeList(map(int, emoji.get('roles', [])))
+        user = emoji.get('user')
+        self.user: Optional[User] = User(state=self._state, data=user) if user else None
+
+    def _to_partial(self) -> PartialEmoji:
+        return PartialEmoji(name=self.name, animated=self.animated, id=self.id)
+
+    def __iter__(self) -> Iterator[Tuple[str, Any]]:
+        for attr in self.__slots__:
+            if attr[0] != '_':
+                value = getattr(self, attr, None)
+                if value is not None:
+                    yield (attr, value)
+
+    def __str__(self) -> str:
+        if self.animated:
+            return f'<a:{self.name}:{self.id}>'
+        return f'<:{self.name}:{self.id}>'
+
+    def __repr__(self) -> str:
+        return f'<{self.__class__.__name__} id={self.id} name={self.name!r} animated={self.animated}>'
+
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, _EmojiTag) and self.id == other.id
+
+    def __ne__(self, other: object) -> bool:
+        return not self.__eq__(other)
+
+    def __hash__(self) -> int:
+        return self.id >> 22
+
+    @property
+    def created_at(self) -> datetime:
+        """:class:`datetime.datetime`: Returns the emoji's creation time in UTC."""
+        return utils.snowflake_time(self.id)
+
+    @property
+    def url(self) -> str:
+        """:class:`str`: Returns the URL of the emoji."""
+        fmt = 'gif' if self.animated else 'png'
+        return f'{Asset.BASE}/emojis/{self.id}.{fmt}'
+
+    @property
+    def roles(self) -> List[Role]:
+        """List[:class:`Role`]: A :class:`list` of roles that is allowed to use this emoji.
+
+        If roles is empty, the emoji is unrestricted.
+        """
+        return []  # there is no way to set roles for app emojis (yet?)
```

Comment:
It's going to be better for this to exist so that places that result in having a mix of app emojis and not (such as reactions on messages) don't end up with people needing to check what kind of emoji before interacting with `.roles`